### PR TITLE
Update page routing, add tests

### PR DESCRIPTION
**Pull request by Nigel2392**
Fix/improvement for issue https://github.com/wagtail/wagtail/issues/11625

* Set `_cached_parent_obj` when a parent `Page` tries to route to a child. (treebeard provided attribute; they set this in the `get_parent` method.)
* Added tests to check if the method exists; I haven't bothered to look at treebeard's docs; very much doubt the attribute is documented. Their API might change - we want to get ahead of that.

### CODE REVIEW

**Comment by Nigel2392** [comment]
hey I notice tests fail on circleCI due to formatting issues; I tried making the formatter happy but it isn't budging. I might go to battle with it. Any ideas how I can defeat the angry formatter?

**Comment by lb-** [comment]
Try `make format-server`

**Comment by Nigel2392** [comment]
> Try `make format-server`
> 
> See also https://docs.wagtail.org/en/stable/contributing/python_guidelines.html#python-coding-guidelines

Perfect! Ruff really likes to live up to it's name. 😉

File: wagtail/tests/test_page_model.py
Comment (ababic) [inline wagtail/tests/test_page_model.py:445]: Since we're only optimising internals here, I was thinking we could probably get away with adding a `self.assertNumQueries()` to an existing test. What do you think @Nigel2392?

File: wagtail/tests/test_page_model.py
Comment (ababic) [inline wagtail/tests/test_page_model.py:3939]: I wonder if this one could be optimised somewhat, for example: just take the root page, set `_cached_parent_obj` to some nonsense variable, then check `get_parent()` returns the nonsense variable. The assertion failure message could even say something super obvious, like:
> Page.get_parent() no longer uses _cached_parent_obj as a cache. We  should probably revise how this is used  to optimise page routing

What do you think?

**Review by ababic — COMMENTED** [review]
Stellar work @Nigel2392! It took a bit of convincing, but this is definitely a sensible improvement. I just had a couple of ideas around tests.

**Comment by lb-** [comment]
@ababic  - did you mind taking another look?

**Comment by lb-** [comment]
@RealOrangeOne  - just pinging you for any quick input / insights.

**Review by ababic — COMMENTED** [review]
I think we could just do with adding an `assertNumQueries` to an existing `route()` test, so that we maintain that baseline. Otherwise I'm happy! I should add that I'm not a member of the core team, so these changes would need a review from somebody else  (possibly @RealOrangeOne) before merging.

**Comment by Nigel2392** [comment]
> I think we could just do with adding an `assertNumQueries` to an existing `route()` test, so that we maintain that baseline. Otherwise I'm happy! I should add that I'm not a member of the core team, so these changes would need a review from somebody else (possibly @RealOrangeOne) before merging.

Ahh sorry! Thought you meant do away with it completely.

Will add later - lot's of things to do right now.

I think the numqueries call can maybe be implied from the value not changing after calling `get_parent`?

**Comment by RealOrangeOne** [comment]
I'm not on the core team either, but I agree that some tests against `route`, particularly using `assertNumQueries` would be great! You could imply it from the return value, but given this is strictly a performance improvement, it'd be good to test the performance directly.

**Comment by Nigel2392** [comment]
Yeah; it's the reason I added this before.

I'll add it back if everyone thinks this is OK

```python
    def test_cached_parent_obj_set(self):
        homepage = Page.objects.get(url_path="/home/")
        christmas_page = EventPage.objects.get(url_path="/home/events/christmas/")

        request = get_dummy_request(path="/events/christmas/")
        (found_page, args, kwargs) = homepage.route(request, ["events", "christmas"])
        self.assertEqual(found_page, christmas_page)

        # parent cache should be set
        events_page = Page.objects.get(url_path="/home/events/").specific
        with self.assertNumQueries(0):
            parent = found_page.get_parent(update=False)
            self.assertEqual(parent, events_page)
```

**Comment by Nigel2392** [comment]
Tests are failing - this doesn't have anything to do with what I wrote; right?

**Comment by gasman** [comment]
The test failure looks like #11804 - if you rebase this branch on top of the current main branch, that should sort it.

**Review by zerolab — APPROVED** [review]
LGTM

**Pull request merged**
